### ISSUE TITLE
Need a failover proxy to localhost/remote server

### ISSUE THREAD

**Issue by OmarMcAdam** [McPizza0/random#1]
TL;DR: Have 1 address `abc.123.com`. when I am running dev server locally, route to me. If not, route to another address.

Long version
UnInbox, has some long running servers that live in a vps and send out webhook messages

I need a proxy with failover capabilities to live permanently online (vps is fine)
when the app is running locally on my machine, it will route to me
when im not, it will route to another web server

I tried with Cloudflare Tunnels and CaddyServer
i can access the tunnel directly using `https://tunnelAddress.mydomain.dev`

CaddyServer installed on a vps and given it the following config
```
proxy.mydomain.dev {
reverse_proxy * {
                to https://tunnelAddress.mydomain.dev https://failover.mydomain.dev
                lb_policy first
                lb_try_duration 5s
                lb_try_interval 250ms

                fail_duration 30s
                max_fails 1
                unhealthy_status 5xx
                unhealthy_latency 2s
                unhealthy_request_count 1
        }
}
```

the path `https://tunnelAddress.mydomain.dev` works fine when accessed directly - validating that the tunnel is up
when accessing via `proxy.mydomain.dev` i get an error `ERR_TOO_MANY_REDIRECTS`

Offering a bounty for a solution that works

**Comment by OmarMcAdam** [comment]
conditions for bounty:

you must provide a working config via PR and explain how it works/what needs to be set up
if you're using cloudflare tunnels, you need to provide details config
if using caddy, provide detailed config

put configs in a pr, I will test, if it works, bounty is yours

**Comment by TheVixhal** [comment]
/attempt #1

<div id="algora" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clsp1bs6r0006l10fu13no3cc/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by TheVixhal** [comment]
#1
/claim #1

**1.Install CaddyServer**: Ensure that CaddyServer is installed on your VPS.

**2.Configure Caddyfile:** Create or modify the Caddyfile with the following configuration:

`proxy.mydomain.dev {
    # Define a matcher for requests coming from localhost
    @local {
        header Host localhost
    }
    
    # Reverse proxy for requests coming from localhost
    reverse_proxy @local {
        to http://localhost:your_local_port
    }

    # Reverse proxy for all other requests
    reverse_proxy {
        # Define the primary upstream server
        to https://tunnelAddress.mydomain.dev
        
        # Define the failover upstream server
        to https://failover.mydomain.dev
    }
}
`
**Explanation:**

**proxy.mydomain.dev:** This directive specifies that this configuration applies to requests made to proxy.mydomain.dev.

**@local:** This is a matcher that checks if the request is coming from localhost by inspecting the Host header. If the Host header contains "localhost", the request is considered to be from localhost.

**reverse_proxy @local { ... }:** This block specifies the reverse proxy settings for requests coming from localhost. It directs such requests to http://localhost:your_local_port, which should be replaced with the actual port number your local development server is running on.

**reverse_proxy { ... }:** This block specifies the reverse proxy settings for all other requests. It defines two upstream servers:

**https://tunnelAddress.mydomain.dev:** This is the primary upstream server. Requests are initially directed here.

**https://failover.mydomain.dev:** This is the failover upstream server. If the primary server is unavailable, requests are automatically rerouted here.

This configuration ensures that requests from localhost are routed to your local development server, while all other requests are initially directed to the primary upstream server. If the primary server fails or becomes unavailable, requests are automatically redirected to the failover server.

**Comment by OmarMcAdam** [comment]
the caddy server lives in a vps, not on my local machine, can you clarify how @local would help?

have you tried this config with a cloudflare tunnel?

the Caddy config was fine, forwarding requests to the cloudflare tunnel
but somewhere between them, there were some issues with too many redirects

https://tunneladdress.mydomain.dev/ will be a cloudflare tunnel into my local system

**Comment by TheVixhal** [comment]
#1


Thank you for the clarification. If CaddyServer is running on a VPS, the @local matcher wouldn't be applicable because it's designed to identify requests originating from localhost, which would be the local machine where CaddyServer isn't running in this case.

For your scenario with Cloudflare Tunnel and CaddyServer on a VPS, you can simplify the configuration to directly proxy requests to the Cloudflare Tunnel address and the failover server. 

```
proxy.mydomain.dev {
    reverse_proxy {
        to https://tunnelAddress.mydomain.dev
        to https://failover.mydomain.dev
    }
}

```

This configuration file instructs CaddyServer to proxy all incoming requests to either https://tunnelAddress.mydomain.dev or https://failover.mydomain.dev based on their availability and health.

Please replace proxy.mydomain.dev, https://tunnelAddress.mydomain.dev, and https://failover.mydomain.dev with your actual domain and addresses respectively.

Once you've updated the configuration file, save it as Caddyfile in your CaddyServer's configuration directory. Then, restart CaddyServer for the changes to take effect.

**Comment by OmarMcAdam** [comment]
Sounds a bit like a GPT

Re-read the issue again, that is how the config file was
the issue is not there
test it yourself

**Comment by TheVixhal** [comment]
> Re-read the issue again, that is how the config file was
> the issue is not there
> test it yourself

share your cadyserver logs.

**Comment by ezhil56x** [comment]
@McPizza0 

- You are running an application on your VPS which is accessible through `abc.123.com`. And when you are running the same on your local machine you need `abc.123.com` to point to your local machine. Is this the expected behaviour?
- And do you have a Static IP or want to make it public with Cloudflare tunnel?
- Is the Caddy server and the application running on same VPS?

**Comment by OmarMcAdam** [comment]
@ezhil56x 
- I am running a service online on VPS A which is accessible through `service.123.com`
- I also occasionally run the same service locally on my machine, which is accessible online at `local.123.com` through a cloudflare tunnel
- Caddy is running on a different vps online, through `caddy.123.com`

I want to be able send webhook messages to `caddy.123.com` and if I am online, it forwards them to `local.123.com`, if I am offline it forwards them to `service.123.com`

**Comment by ezhil56x** [comment]
@McPizza0

Thanks for the info. It would be easier to troubleshoot if we could schedule an online meet

**Comment by OmarMcAdam** [comment]
its easy enough to replicate the setup
dont need a meeting for that

im not looking for "technical support" and diagnosis, i need a solution that works

**Comment by TheVixhal** [comment]
> @ezhil56x
> 
> * I am running a service online on VPS A which is accessible through `service.123.com`
> * I also occasionally run the same service locally on my machine, which is accessible online at `local.123.com` through a cloudflare tunnel
> * Caddy is running on a different vps online, through `caddy.123.com`
> 
> I want to be able send webhook messages to `caddy.123.com` and if I am online, it forwards them to `local.123.com`, if I am offline it forwards them to `service.123.com`

```
caddy.123.com {

    # Define a route for handling errors
    handle_errors {
    
        # Define a condition for when to trigger this route
        @offline {
            expression {http.error.status_code} == 502 # 502 Bad Gateway
        }
        
        # Proxy requests to service.123.com when offline
        reverse_proxy @offline https://service.123.com
    }

    # Define a route for normal operation
    reverse_proxy https://local.123.com {
    
        # Set the Host header to the upstream host
        header_up Host {http.reverse_proxy.upstream.hostport}
    }
}

```

Replace caddy.123.com, local.123.com, and service.123.com with your actual domain names.

**Comment by OmarMcAdam** [comment]
@TheVixhal 
in theory, should work great, nice directive with the `handle_errors`

In practice, it dosnt work
when the tunnel is down, initially it gives a 504, but the proxy condition didnt work
then it switched to a different error
![image](https://github.com/McPizza0/random/assets/17185737/5b6b8be1-775d-447f-ac03-b21b06c25807)

so it seems that the tunnel always responds with something

this may not be viable after all
would need to add a way for it to check if the host is up by hitting a `/health` endpoint and only forwarding if it gets back a 200 code

**Comment by OmarMcAdam** [comment]
Thanks for trying @ezhil56x and @TheVixhal 

closing this issue and removing the bounty
already spent 6 hours on this

will build out something better in less time: https://github.com/uninbox/webhook-proxy

**Comment by francislavoie** [comment]
The answer is in the docs: https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#https

**Comment by TheVixhal** [comment]
> @TheVixhal in theory, should work great, nice directive with the `handle_errors`
> 
> In practice, it dosnt work when the tunnel is down, initially it gives a 504, but the proxy condition didnt work then it switched to a different error ![image](https://private-user-images.githubusercontent.com/17185737/305645052-5b6b8be1-775d-447f-ac03-b21b06c25807.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyMDM3ODMsIm5iZiI6MTcwODIwMzQ4MywicGF0aCI6Ii8xNzE4NTczNy8zMDU2NDUwNTItNWI2YjhiZTEtNzc1ZC00NDdmLWFjMDMtYjIxYjA2YzI1ODA3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE3VDIwNTgwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmYjU4MjVkZjljMzM2YTljM2Q4Y2RlZDk0ZWRlNDZiYzg4NjZmMWM1ZTBkNjM2NGYxNDQ3ZTU4ZDhkMTVmNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.a-4rdwMRkuVaCKd5mXDS1svpD5Eh2wZGQbjuBAAOrm8)
> 
> so it seems that the tunnel always responds with something
> 
> this may not be viable after all would need to add a way for it to check if the host is up by hitting a `/health` endpoint and only forwarding if it gets back a 200 code

```
caddy.123.com {

    # Use trusted TLS certificates from a certificate authority
    tls internal

    # Enhanced health checks for better reliability
    handle_path /health {
        reverse_proxy https://local.123.com/health {
            health_check {
                interval 30s
                timeouts 5s
                failures 3
                passes 2
            }
        }
    }

    # Online routing to the local server with load balancing
    @online {
        expression {upstream_status} == "healthy"
    }
    reverse_proxy @online {
        lb_policy least_conn # Load balancing policy changed to least connections
        lb_try_duration 1s # Retry duration for failed attempts
        https://local.123.com
    }

    # Offline routing to the service server
    reverse_proxy https://service.123.com

    # Custom error pages with enhanced content and appearance
    handle_errors {
        root * /path/to/custom/error/pages
        file_server
        status 404 /404.html # Custom handling for 404 errors
    }

    # Logging configuration with rotating log files
    log {
        output file /var/log/caddy/access.log {
            roll_size 5MB # Rotate logs at 5MB
            roll_keep 7 # Keep 7 days of logs
            roll_keep_for 30d # Keep logs for 30 days
        }
        format json
    }

    # Automatic HTTPS redirection for all traffic
    redir {
        if {scheme} is http
        / https://{host}{uri}
    }

    # Additional metrics for more detailed server information
    metrics {
        prometheus 0.0.0.0:9180
        # Add more metrics if needed
    }

    # Caching configuration for static assets
    route /static* {
        handle {
            root * /path/to/static/assets
            file_server
            header {
                Cache-Control "public, max-age=604800" # Example cache control header
            }
        }
    }

    # Security headers with enhanced content security policy
    handle {
        header {
            Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; report-uri /csp-report-endpoint;" # Enhanced CSP with reporting
            Strict-Transport-Security "max-age=63072000; includeSubDomains; preload" # Example HSTS header
            X-Content-Type-Options "nosniff" # Example X-Content-Type-Options header
            X-Frame-Options "DENY" # Added X-Frame-Options header for clickjacking protection
            Referrer-Policy "strict-origin-when-cross-origin" # Added Referrer-Policy header for improved privacy
            Feature-Policy "camera 'none'; microphone 'none'" # Added Feature-Policy header to restrict access to certain features
        }
    }

    # Rate limiting and DDoS protection with enhanced configuration
    handle_errors {
        ratelimit {
            rule 5s allow 2 # Example rate limiting rule
            error 429 "Rate limit exceeded, please try again later."
        }
    }
}

```

**Comment by TheVixhal** [comment]
> > @TheVixhal in theory, should work great, nice directive with the `handle_errors`
> > In practice, it dosnt work when the tunnel is down, initially it gives a 504, but the proxy condition didnt work then it switched to a different error ![image](https://private-user-images.githubusercontent.com/17185737/305645052-5b6b8be1-775d-447f-ac03-b21b06c25807.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyMDM3ODMsIm5iZiI6MTcwODIwMzQ4MywicGF0aCI6Ii8xNzE4NTczNy8zMDU2NDUwNTItNWI2YjhiZTEtNzc1ZC00NDdmLWFjMDMtYjIxYjA2YzI1ODA3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE3VDIwNTgwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmYjU4MjVkZjljMzM2YTljM2Q4Y2RlZDk0ZWRlNDZiYzg4NjZmMWM1ZTBkNjM2NGYxNDQ3ZTU4ZDhkMTVmNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.a-4rdwMRkuVaCKd5mXDS1svpD5Eh2wZGQbjuBAAOrm8)
> > so it seems that the tunnel always responds with something
> > this may not be viable after all would need to add a way for it to check if the host is up by hitting a `/health` endpoint and only forwarding if it gets back a 200 code
> 
> ```
> caddy.123.com {
> 
>     # Use trusted TLS certificates from a certificate authority
>     tls internal
> 
>     # Enhanced health checks for better reliability
>     handle_path /health {
>         reverse_proxy https://local.123.com/health {
>             health_check {
>                 interval 30s
>                 timeouts 5s
>                 failures 3
>                 passes 2
>             }
>         }
>     }
> 
>     # Online routing to the local server with load balancing
>     @online {
>         expression {upstream_status} == "healthy"
>     }
>     reverse_proxy @online {
>         lb_policy least_conn # Load balancing policy changed to least connections
>         lb_try_duration 1s # Retry duration for failed attempts
>         https://local.123.com
>     }
> 
>     # Offline routing to the service server
>     reverse_proxy https://service.123.com
> 
>     # Custom error pages with enhanced content and appearance
>     handle_errors {
>         root * /path/to/custom/error/pages
>         file_server
>         status 404 /404.html # Custom handling for 404 errors
>     }
> 
>     # Logging configuration with rotating log files
>     log {
>         output file /var/log/caddy/access.log {
>             roll_size 5MB # Rotate logs at 5MB
>             roll_keep 7 # Keep 7 days of logs
>             roll_keep_for 30d # Keep logs for 30 days
>         }
>         format json
>     }
> 
>     # Automatic HTTPS redirection for all traffic
>     redir {
>         if {scheme} is http
>         / https://{host}{uri}
>     }
> 
>     # Additional metrics for more detailed server information
>     metrics {
>         prometheus 0.0.0.0:9180
>         # Add more metrics if needed
>     }
> 
>     # Caching configuration for static assets
>     route /static* {
>         handle {
>             root * /path/to/static/assets
>             file_server
>             header {
>                 Cache-Control "public, max-age=604800" # Example cache control header
>             }
>         }
>     }
> 
>     # Security headers with enhanced content security policy
>     handle {
>         header {
>             Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; report-uri /csp-report-endpoint;" # Enhanced CSP with reporting
>             Strict-Transport-Security "max-age=63072000; includeSubDomains; preload" # Example HSTS header
>             X-Content-Type-Options "nosniff" # Example X-Content-Type-Options header
>             X-Frame-Options "DENY" # Added X-Frame-Options header for clickjacking protection
>             Referrer-Policy "strict-origin-when-cross-origin" # Added Referrer-Policy header for improved privacy
>             Feature-Policy "camera 'none'; microphone 'none'" # Added Feature-Policy header to restrict access to certain features
>         }
>     }
> 
>     # Rate limiting and DDoS protection with enhanced configuration
>     handle_errors {
>         ratelimit {
>             rule 5s allow 2 # Example rate limiting rule
>             error 429 "Rate limit exceeded, please try again later."
>         }
>     }
> }
> ```

Customise according to your need😁

**Comment by francislavoie** [comment]
@TheVixhal stop using ChatGPT. That config is complete nonsense.

Source: I'm a Caddy maintainer. I wrote most of the Caddyfile docs.

**Comment by TheVixhal** [comment]
> @TheVixhal stop using ChatGPT. That config is complete nonsense.
Sorry 😔

**Comment by TheVixhal** [comment]
> @TheVixhal stop using ChatGPT. That config is complete nonsense.
> 
> Source: I'm a Caddy maintainer. I wrote most of the Caddyfile docs.

Sorry 😔😔😐
